### PR TITLE
add watcher for expiration, add expiration for TestAndSet

### PR DESCRIPTION
**Pull request by yifan-gu**
also add some test cases

### CODE REVIEW

**Comment by xiang90** [comment]
@evan-gu 
WatcherHub should keep the last `Index` and `Term`. So you do not need to calculate by finding the last item in the queue, which is not quite readable. 
When we have duplicate `index` in the watcherHub, the scan start index of the queue is not `index` - `startIndex`.

In function `addEvent(e *Event)`, you can check the event's `index`. If the `index` is 0, you should use the last known index in the history. Having a new function only add a little logic is not good.

Comments the code, not in github.

**Comment by yifan-gu** [comment]
ok

**Comment by yifan-gu** [comment]
I am going to sleep. I will fix it tomorrow.

"When we have duplicate index in the watcherHub, the scan start index of the queue is not index - startIndex"

what does that mean?

**Comment by yifan-gu** [comment]
@xiangli-cmu

File: store/event.go [hunk 1]
```diff
@@ -102,15 +105,43 @@ func newEventHistory(capacity int) *EventHistory {
...
 	eh.StartIndex = eh.Queue.Events[eh.Queue.Front].Index
+
+	eh.LastIndex = e.Index
+	eh.LastTerm = e.Term
+
+	return e
 }
 
+// addEvent with the last event's index and term
+/*func (eh *EventHistory) addEventWithouIndex(action, key string) (e *Event) {
```
Comment (xiang90) [inline store/event.go]: Hmm... is this function still needed?

File: store/node.go [hunk 2]
```diff
@@ -265,7 +265,14 @@ func (n *Node) Expire() {
 		select {
 		// if timeout, delete the node
 		case <-time.After(duration):
+			e := newEvent(Expire, n.Path, 0, 0)
```
Comment (xiang90) [inline store/node.go]: Can you define the two zero?
Like `undefinedInde` and `undefinedTerm`

File: store/event.go [hunk 3]
```diff
@@ -89,6 +92,9 @@ func (eq *eventQueue) insert(e *Event) {
 type EventHistory struct {
 	Queue      eventQueue
 	StartIndex uint64
+	LastIndex  uint64
+	LastTerm   uint64
+	DupIndex   uint64
```
Comment (xiang90) [inline store/event.go]: DupCnt?

File: store/event.go [hunk 4]
```diff
@@ -102,13 +108,31 @@ func newEventHistory(capacity int) *EventHistory {
...
-func (eh *EventHistory) addEvent(e *Event) {
+func (eh *EventHistory) addEvent(e *Event) *Event {
 	eh.rwl.Lock()
 	defer eh.rwl.Unlock()
 
+	DupIndex := uint64(0)
+
+	if e.Index == UndefIndex {
+		e.Index = eh.LastIndex
+		DupIndex = 1
```
Comment (xiang90) [inline store/event.go]: duped?

**Pull request merged**